src/libstd/sys/cloudabi/io.rs

```diff
@@ -20,6 +27,13 @@ impl<'a> IoSliceMut<'a> {
         IoSliceMut(buf)
     }
 
+    #[inline]
+    pub unsafe fn advance(&mut self, n: usize) {
+        // We create a new slice unsafely to get around lifetime issues. Since
+        // the slice need to life at least as long lifetime 'a, this is safe.
+        self.0 = slice::from_raw_parts_mut(self.0.as_mut_ptr().add(n), self.0.len() - n)
```

Comment:
Why is this necessary? It should be possible to "move" the slice along without unsafe code, perhaps [`split_at_mut`](https://doc.rust-lang.org/nightly/std/primitive.slice.html) would be of help here?

Reply:
As the comment says it's to work around lifetime issue, see this [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff0c3cc39ff05c315a511911ff58a4b9).

Reply:
FWIW I personally agree with @Mark-Simulacrum here, this (and the other impls) should be doable without unsafe code